R/rows.R

```diff
@@ -0,0 +1,173 @@
+  if (inherits(..f, "formula")) {
+    ..f <- as_function(..f)
+  } else if (!is.function(..f)) {
+    stop("..f should be a function or a formula", call. = FALSE)
+  }
+
+  if (!dplyr::is.grouped_df(.d)) {
+    ..f(.d, ...)
+  } else {
+    dots <- substitute(pairlist(...))
+    calling_env <- parent.frame()
+    by_slice_impl(.d, ..f, dots, .trace, calling_env)
+  }
+}
+
+
+#' Apply a function to each row of a data frame
+#'
+#' Applies \code{..f} to each row of \code{.d}. By default, the whole
+#' row is appended to the result to serve as identifier (set
+#' \code{.trace} to \code{FALSE} to prevent this). In addition, if
+#' \code{..f} returns a multi-rows data frame or a non-scalar atomic
+#' vector, a \code{.row} column is appended to identify the row number
+#' in the original data frame. If \code{..f}'s output is not a data
+#' frame nor an atomic vector, a list-column is created. In all cases,
+#' \code{by_row()} and \code{map_row()} create a data frame in tidy
+#' format.
+#'
+#' \code{map_rows()} is intended to provide a version of
+#' \code{map_n()} that works better with data frames. The distinction
+#' between \code{by_row()} and \code{map_rows()} is that in the former
+#' case, \code{..f} gets the data frame as argument, while in the
+#' latter, the columns are mapped to \code{..f}'s function
+#' signature. In essence, \code{map_row()} is equivalent to using
+#' \code{by_row()} with a stitched function (see \code{stitch()}).
+#' @param .d A data frame.
+#' @param .f,..f A function to apply to each row. If \code{..f} does
+#'   not return a data frame or an atomic vector, a list-column is
+#'   created under the name \code{.out}. If it returns a data frame, it
+#'   should have the same number of rows within groups and the same
+#'   number of columns between groups.
+#' @param ... Further arguments passed to \code{..f}.
+#' @param .trace If \code{TRUE}, the returned data frame is prepended
+#'   with the columns defining the slices in \code{.d}. They are
+#'   recycled to match the output size in each slice if necessary.
+#' @return A data frame.
+#' @seealso \code{\link{by_slice}()}
+#' @useDynLib purrr
+#' @export
+#' @examples
+#' # ..f should be able to work with a list or a data frame. As it
+#' # happens, sum() handles data frame so the following works:
+#' mtcars %>% by_row(sum)
+#'
+#' \dontrun{ % temporary examples until lift() and punctuate_v() make it (or not) in purrr
+#' # Other functions such as mean() may need to be adjusted:
+#' mtcars %>% by_row(lift(mean))
+#'
+#' # To run a function with map_rows(), make sure it is variadic (that
+#' # it accepts dots)
+#' mtcars %>% map_rows(sum)
+#' mtcars %>% map_rows(punctuate_v(mean))
+#' }
+by_row <- function(.d, ..f, ..., .trace = TRUE) {
+  if (inherits(..f, "formula")) {
+    ..f <- as_function(..f)
+  } else if (!is.function(..f)) {
+    stop("..f should be a function or a formula", call. = FALSE)
+  }
+
+  dots <- substitute(list(...))
+  calling_env <- parent.frame()
```

Comment:
I think that here we should use the `calling_env()` function that I've proposed in a magrittr PR. This way, we'd evaluate `..f` in the original calling environment, even within a pipe chain. This can be important if `..f` has side effects for example.